src/generic_datatype.cpp

```diff
@@ -0,0 +1,175 @@
+
+namespace smt {
+
+GenericDatatypeDecl::GenericDatatypeDecl(const std::string name) : dt_name(name)
+{
+}
+
+std::string GenericDatatypeDecl::get_name() { return dt_name; }
+
+GenericDatatypeConstructorDecl::GenericDatatypeConstructorDecl(
+    const std::string & name)
+    : cons_name(name)
+{
+}
+
+void GenericDatatypeConstructorDecl::add_new_selector(
+    const SelectorComponents & newSelector)
+{
+  for (unsigned int i = 0; i < selector_vector.size(); ++i)
+  {
+    // Checks if the selector has already been added
+    if (selector_vector[i].name == (newSelector).name)
+    {
+      throw "Can't add selector. It already exists in this datatype!";
+    }
+  }
+  selector_vector.push_back(newSelector);
+}
+
+std::vector<SelectorComponents>
+GenericDatatypeConstructorDecl::get_selector_vector()
+{
+  return selector_vector;
+}
+
+std::string GenericDatatypeConstructorDecl::get_name() const
+{
+  return cons_name;
+}
+
+int GenericDatatypeConstructorDecl::get_selector_count() const
+{
+  return selector_vector.size();
+}
+bool GenericDatatypeConstructorDecl::compare(
+    const DatatypeConstructorDecl & d) const
+{
+  // COmpares based off constructor's name
+  return cons_name
+         == static_pointer_cast<GenericDatatypeConstructorDecl>(d)->get_name();
+}
+
+std::string GenericDatatypeConstructorDecl::get_dt_name() const
+{
+  return static_pointer_cast<GenericDatatypeDecl>(dt_decl)->get_name();
+}
+
+void GenericDatatypeConstructorDecl::update_stored_dt(
+    const DatatypeDecl & datatype_decl)
+{
+  dt_decl = datatype_decl;
+}
+
+GenericDatatype::GenericDatatype(const DatatypeDecl & dt_declaration)
+    : dt_decl(dt_declaration)
+{
+}
+
+void GenericDatatype::add_constructor(
+    const DatatypeConstructorDecl & dt_cons_decl)
+{
+  // checks if dt_cons_decl is already associated with the datatype
+  if (std::find(cons_decl_vector.begin(), cons_decl_vector.end(), dt_cons_decl)
+      != cons_decl_vector.end())
+  {
+    throw "Can't add constructor. It already has been added!";
+  }
+  shared_ptr<GenericDatatypeConstructorDecl> gdt_cons =
+      static_pointer_cast<GenericDatatypeConstructorDecl>(dt_cons_decl);
+  // Links the constructor to the datatype_decl of the datatype
+  gdt_cons->update_stored_dt(dt_decl);
+  // Links the datatype to the new constructor
+  cons_decl_vector.push_back(dt_cons_decl);
+}
+
+void GenericDatatype::add_selector(const DatatypeConstructorDecl & dt_cons_decl,
+                                   const SelectorComponents & newSelector)
+{
+  for (unsigned int i = 0; i < cons_decl_vector.size(); ++i)
+  {
+    // If the constructor is associated with the datatype
+    if (cons_decl_vector[i] == dt_cons_decl)
+    {
+      // Adds the selector to the correct constructor
+      static_pointer_cast<GenericDatatypeConstructorDecl>(cons_decl_vector[i])
+          ->add_new_selector(newSelector);
+    }
+    else
+    {
+      throw "Can't add selector. The constructor is not a member of the datatype!";
```

Comment:
Great! Please add a test that makes sure that this exception is thrown for some case where it needs to.

Reply:
Great catch :)